### PR TITLE
Fix Concurrency Thread Group very high CPU issue

### PR DESCRIPTION
**Pull request by pmouawad**
As reported in:

- https://groups.google.com/forum/#!topic/jmeter-plugins/biuDmDSprFk

Improvements on CPU for same test:
Before : Avg:14.5%, Max : 99.1%
After : Avg : 8.53%, Max : 37%

If we zoom on first Minute:
Before : Avg:40%, Max : 99.1%
After : Avg : 17%, Max : 37%

### CODE REVIEW

File: plugins/casutg/src/main/java/com/blazemeter/jmeter/threads/concurrency/ConcurrencyThreadStarter.java [hunk 1]
```diff
@@ -1,31 +1,49 @@
...
 import org.apache.jorphan.collections.ListedHashTree;
 import org.apache.jorphan.logging.LoggingManager;
 import org.apache.log.Logger;
 
+import com.blazemeter.jmeter.threads.AbstractThreadStarter;
+import com.blazemeter.jmeter.threads.DynamicThread;
+
 public class ConcurrencyThreadStarter extends AbstractThreadStarter {
     private static final Logger log = LoggingManager.getLoggerForClass();
+    private static final long DEFAULT_SHIFT_RAMPUP = JMeterUtils.getPropDefault("dynamic_tg.shift_rampup_start", 0L);
```
Comment (undera) [inline plugins/casutg/src/main/java/com/blazemeter/jmeter/threads/concurrency/ConcurrencyThreadStarter.java]: This is one of features to be able to change it in runtime via property. It can't be static, but can be cached like `rampUp`.

File: plugins/casutg/src/main/java/com/blazemeter/jmeter/threads/concurrency/ConcurrencyThreadStarter.java [hunk 2]
```diff
@@ -53,14 +70,32 @@ private long getPlannedConcurrency() {
...
             double stepLen = rampUp / (double) steps;
             return Math.round(stepSize * (Math.floor(timeOffset / stepLen) + 1));
         } else {
-            double slope = maxConcurr / rampUp;
+            double slope = (double) maxConcurr / rampUp;
             return Math.round(slope * timeOffset);
         }
     }
+
+    void checkNeedsPropertiesReloading(long now) {
```
Comment (undera) [inline plugins/casutg/src/main/java/com/blazemeter/jmeter/threads/concurrency/ConcurrencyThreadStarter.java]: Maybe make it private and call in constructor to avoid code duplication?

File: plugins/casutg/src/main/java/com/blazemeter/jmeter/threads/concurrency/ConcurrencyThreadStarter.java [hunk 3]
```diff
@@ -53,14 +70,32 @@ private long getPlannedConcurrency() {
...
+            this.steps = owner.getStepsAsLong();
+            this.maxConcurr = Math.round(owner.getTargetLevelAsDouble());
+            this.lastCachedTime = System.currentTimeMillis();            
+        }
+    }
+
+    /**
+     * @return the lastCachedTime
+     */
+    long getLastCachedTime() {
```
Comment (undera) [inline plugins/casutg/src/main/java/com/blazemeter/jmeter/threads/concurrency/ConcurrencyThreadStarter.java]: If this is for testing needs, it should have `protected` access and be exposed via `ConcurrencyThreadGroupExt` class

Comment (undera) [reply plugins/casutg/src/main/java/com/blazemeter/jmeter/threads/concurrency/ConcurrencyThreadStarter.java]: or have `lastCachedTime` as protected and move this method to `ConcurrencyThreadGroupExt`

File: plugins/casutg/src/test/java/com/blazemeter/jmeter/threads/concurrency/ConcurrencyThreadGroupTest.java [hunk 4]
```diff
@@ -104,9 +107,24 @@ public void testStartNextLoop() throws Exception {
...
+    @Test
+    public void testCachingOfProperties() throws InterruptedException {
+        Object[] objects = createTestPlan();
+        ListedHashTree hashTree = (ListedHashTree) objects[0];
+        ConcurrencyThreadGroupExt ctg = (ConcurrencyThreadGroupExt) objects[1];
+        
+        ConcurrencyThreadStarter starter = new ConcurrencyThreadStarter(0, new ListenerNotifier(), 
+                hashTree, new StandardJMeterEngine(), ctg);
+        long lastCachedTime = starter.getLastCachedTime();
+        Thread.sleep(ConcurrencyThreadStarter.CACHING_VALIDITY_MS/2);
```
Comment (undera) [inline plugins/casutg/src/test/java/com/blazemeter/jmeter/threads/concurrency/ConcurrencyThreadGroupTest.java]: please reformat the code so there are spaces around `/` and `*` in equasions

File: plugins/casutg/src/main/java/com/blazemeter/jmeter/threads/AbstractThreadStarter.java [hunk 5]
```diff
@@ -44,11 +44,12 @@ public void run() {
             supplyActiveThreads();
         } catch (InterruptedException e) {
             log.debug("Interrupted", e);
+            Thread.currentThread().interrupt();
```
Comment (undera) [inline plugins/casutg/src/main/java/com/blazemeter/jmeter/threads/AbstractThreadStarter.java]: Why is this needed? I assume the `run()` method will just finish and corresponding thread will finish.

File: plugins/casutg/src/main/java/com/blazemeter/jmeter/threads/concurrency/ConcurrencyThreadGroup.java [hunk 6]
```diff
@@ -1,32 +1,51 @@
...
 import org.apache.jorphan.collections.ListedHashTree;
 import org.apache.jorphan.logging.LoggingManager;
 import org.apache.log.Logger;
 
+import com.blazemeter.jmeter.threads.AbstractDynamicThreadGroup;
+
 public class ConcurrencyThreadGroup extends AbstractDynamicThreadGroup {
     private static final Logger log = LoggingManager.getLoggerForClass();
-    public static final int MIN_CHECK_TIME = 1000;
+    private static final long DEFAULT_TEMPORISATION = JMeterUtils.getPropDefault("dynamic_tg.temporisation", 10L);
```
Comment (undera) [inline plugins/casutg/src/main/java/com/blazemeter/jmeter/threads/concurrency/ConcurrencyThreadGroup.java]: This property will need a note in documentation, so users and us will know it exists for the future.

File: plugins/casutg/src/main/java/com/blazemeter/jmeter/threads/concurrency/ConcurrencyThreadStarter.java [hunk 7]
```diff
@@ -53,14 +70,32 @@ private long getPlannedConcurrency() {
...
             return Math.round(slope * timeOffset);
         }
     }
+
+    void checkNeedsPropertiesReloading(long now) {
+        if (CACHING_VALIDITY_MS > 0 && now - lastCachedTime > CACHING_VALIDITY_MS) {
+            this.rampUp = owner.getRampUpSeconds();
+            this.hold = owner.getHoldSeconds();
+            this.steps = owner.getStepsAsLong();
+            this.maxConcurr = Math.round(owner.getTargetLevelAsDouble());
```
Comment (undera) [inline plugins/casutg/src/main/java/com/blazemeter/jmeter/threads/concurrency/ConcurrencyThreadStarter.java]: original algorighm has no rounding. I'd avoid introducing it unless there is good reason.

File: plugins/casutg/src/main/java/com/blazemeter/jmeter/threads/AbstractDynamicThreadGroupModel.java [hunk 8]
```diff
@@ -24,7 +24,7 @@
     public static final String STEPS = "Steps";
     public static final String ITERATIONS = "Iterations";
     public static final String HOLD = "Hold";
-    protected final Set<DynamicThread> threads = Collections.newSetFromMap(new ConcurrentHashMap<DynamicThread, Boolean>());
+    protected transient Set<DynamicThread> threads = Collections.newSetFromMap(new ConcurrentHashMap<DynamicThread, Boolean>());
```
Comment (undera) [inline plugins/casutg/src/main/java/com/blazemeter/jmeter/threads/AbstractDynamicThreadGroupModel.java]: why remove `final`?

**Comment by undera** [comment]
The general approach is good. Please address my minor comments.

File: plugins/casutg/src/main/java/com/blazemeter/jmeter/threads/AbstractDynamicThreadGroupModel.java (hunk 8 above)
Comment (pmouawad) [reply plugins/casutg/src/main/java/com/blazemeter/jmeter/threads/AbstractDynamicThreadGroupModel.java]: transient cannot be final  IMU : http://blog.clempinch.com/transient-and-final-instance-variables-in-java/

File: plugins/casutg/src/main/java/com/blazemeter/jmeter/threads/concurrency/ConcurrencyThreadStarter.java (hunk 7 above)
Comment (pmouawad) [reply plugins/casutg/src/main/java/com/blazemeter/jmeter/threads/concurrency/ConcurrencyThreadStarter.java]: getTargetLevelAsDouble() transforms an int into a double. I do it once to avoid redoing it later each time

File: plugins/casutg/src/main/java/com/blazemeter/jmeter/threads/concurrency/ConcurrencyThreadGroup.java (hunk 6 above)
Comment (pmouawad) [reply plugins/casutg/src/main/java/com/blazemeter/jmeter/threads/concurrency/ConcurrencyThreadGroup.java]: I didn't find documentation for the other one "dynamic_tg.shift_rampup_start", in which file should it be documented ?

File: plugins/casutg/src/main/java/com/blazemeter/jmeter/threads/AbstractThreadStarter.java (hunk 5 above)
Comment (pmouawad) [reply plugins/casutg/src/main/java/com/blazemeter/jmeter/threads/AbstractThreadStarter.java]: AFAIU :  "InterruptedException" should not be ignored (squid:S2142)

File: plugins/casutg/src/test/java/com/blazemeter/jmeter/threads/concurrency/ConcurrencyThreadGroupTest.java (hunk 4 above)
Comment (pmouawad) [reply plugins/casutg/src/test/java/com/blazemeter/jmeter/threads/concurrency/ConcurrencyThreadGroupTest.java]: Will be in next commit

File: plugins/casutg/src/main/java/com/blazemeter/jmeter/threads/concurrency/ConcurrencyThreadStarter.java (hunk 1 above)
Comment (pmouawad) [reply plugins/casutg/src/main/java/com/blazemeter/jmeter/threads/concurrency/ConcurrencyThreadStarter.java]: Will be in next commit

File: plugins/casutg/src/main/java/com/blazemeter/jmeter/threads/concurrency/ConcurrencyThreadStarter.java (hunk 2 above)
Comment (pmouawad) [reply plugins/casutg/src/main/java/com/blazemeter/jmeter/threads/concurrency/ConcurrencyThreadStarter.java]: As you guessed later for JUNIT

File: plugins/casutg/src/main/java/com/blazemeter/jmeter/threads/concurrency/ConcurrencyThreadStarter.java (hunk 3 above)
Comment (pmouawad) [reply plugins/casutg/src/main/java/com/blazemeter/jmeter/threads/concurrency/ConcurrencyThreadStarter.java]: I don't understand, ConcurrencyThreadGroupExt does not extend ConcurrencyThreadStarter.
Can you clarify ? 
Thanks

File: plugins/casutg/src/main/java/com/blazemeter/jmeter/threads/AbstractThreadStarter.java (hunk 5 above)
Comment (undera) [reply plugins/casutg/src/main/java/com/blazemeter/jmeter/threads/AbstractThreadStarter.java]: Sorry, I don't understand this reason. It is not ignored, it is logged. But the situation does not require any action apart from logging, the thread will just finish its job naturally.

File: plugins/casutg/src/main/java/com/blazemeter/jmeter/threads/concurrency/ConcurrencyThreadGroup.java (hunk 6 above)
Comment (undera) [reply plugins/casutg/src/main/java/com/blazemeter/jmeter/threads/concurrency/ConcurrencyThreadGroup.java]: It's this file: https://github.com/undera/jmeter-plugins/blob/master/site/dat/wiki/ConcurrencyThreadGroup.wiki
I'll add there other relevant properties after merging this PR.

File: plugins/casutg/src/main/java/com/blazemeter/jmeter/threads/AbstractThreadStarter.java (hunk 5 above)
Comment (pmouawad) [reply plugins/casutg/src/main/java/com/blazemeter/jmeter/threads/AbstractThreadStarter.java]: This is what the squid:S2142 says: IMU we must interrupt to be correct:

> InterruptedExceptions should never be ignored in the code, and simply logging the exception counts in this case as "ignoring". The throwing of the InterruptedException clears the interrupted state of the Thread, so if the exception is not handled properly the fact that the thread was interrupted will be lost. Instead, InterruptedExceptions should either be rethrown - immediately or after cleaning up the method's state - or the thread should be re-interrupted by calling Thread.interrupt() even if this is supposed to be a single-threaded application. Any other course of action risks delaying thread shutdown and loses the information that the thread was interrupted - probably without finishing its task.

File: plugins/casutg/src/main/java/com/blazemeter/jmeter/threads/concurrency/ConcurrencyThreadStarter.java (hunk 7 above)
Comment (undera) [reply plugins/casutg/src/main/java/com/blazemeter/jmeter/threads/concurrency/ConcurrencyThreadStarter.java]: This can change the math afterwards. Can we keep the math, unless we have a bug there?
These changes affect so wide audience that I want to avoid changes as much as possible.

File: plugins/casutg/src/main/java/com/blazemeter/jmeter/threads/concurrency/ConcurrencyThreadStarter.java (hunk 3 above)
Comment (undera) [reply plugins/casutg/src/main/java/com/blazemeter/jmeter/threads/concurrency/ConcurrencyThreadStarter.java]: Probably I'm wrong and misunderstood the change. My bad

File: plugins/casutg/src/main/java/com/blazemeter/jmeter/threads/AbstractThreadStarter.java (hunk 5 above)
Comment (undera) [reply plugins/casutg/src/main/java/com/blazemeter/jmeter/threads/AbstractThreadStarter.java]: Still I don't see how doing interrupt changes anything... The code is so small and simple, it has not finish the thread...
Ok, let's keep it for now.

File: plugins/casutg/src/main/java/com/blazemeter/jmeter/threads/concurrency/ConcurrencyThreadStarter.java (hunk 7 above)
Comment (pmouawad) [reply plugins/casutg/src/main/java/com/blazemeter/jmeter/threads/concurrency/ConcurrencyThreadStarter.java]: In fact I do the same as before but I cache. Do you want me to revert ?

Comment (undera) [reply plugins/casutg/src/main/java/com/blazemeter/jmeter/threads/concurrency/ConcurrencyThreadStarter.java]: Yes, please

File: plugins/casutg/src/main/java/com/blazemeter/jmeter/threads/concurrency/ConcurrencyThreadGroup.java (hunk 6 above)
Comment (undera) [inline plugins/casutg/src/main/java/com/blazemeter/jmeter/threads/concurrency/ConcurrencyThreadGroup.java]: This property also needs documenting

**Comment by undera** [comment]
Thanks for all the fixes. Now I have much more confidence merging it.
Thanks for your patience!

BTW I will release the new version right away, but it won't make it to appear on BlazeMeter engines, because this plugin is pre-installed on engine images...

**Comment by ping0uin** [comment]
Hi @undera ,
do you know when it will be available on the blazemeter engines? This is a high priority for us. We cannot use properly, and spread the solution in our company without this...
Thanks.

**Comment by undera** [comment]
I have no idea when it will get into BlazeMeter. Try asking their support.
